### PR TITLE
fix many "Only booleans are allowed in an if condition"

### PR DESCRIPTION
**Pull request by ColonelMoutarde**
Changes proposed in this pull request:

- By resetting the "booleansInConditions" directive to true, 370 errors are detected. I've corrected 20 of the errors that phpstan is reporting. There are still many...

Pull request checklist:

- [x] clear commit messages
- [x] code manually tested

### CODE REVIEW

File: cli/create-user.php [hunk 1]
```diff
@@ -10,7 +10,7 @@
 }
 
 $usernames = listUsers();
-if (preg_grep("/^$username$/i", $usernames)) {
+if (is_bool(preg_grep("/^$username$/i", $usernames))) {
```
Comment (Alkarex) [inline cli/create-user.php]: I would prefer:

```suggestion
if (preg_grep("/^$username$/i", $usernames) !== false) {
```

and likewise for other similar cases below

File: cli/i18n/I18nUsageValidator.php [hunk 2]
```diff
@@ -42,7 +42,7 @@ public function validate(): bool {
 		foreach ($this->reference as $file => $data) {
 			foreach ($data as $key => $value) {
 				$this->totalEntries++;
-				if (preg_match('/\._$/', $key) && in_array(preg_replace('/\._$/', '', $key), $this->code, true)) {
+				if (is_int(preg_match('/\._$/', $key)) && in_array(preg_replace('/\._$/', '', $key), $this->code, true)) {
```
Comment (Alkarex) [inline cli/i18n/I18nUsageValidator.php]: I would prefer something like the following (or `!= false`)

```suggestion
				if (preg_match('/\._$/', $key) === 1 && in_array(preg_replace('/\._$/', '', $key), $this->code, true)) {
```

File: p/api/greader.php [hunk 3]
```diff
@@ -1069,7 +1069,7 @@ public static function parse() {
 								$include_target = $pathInfos[7];
 								if ($include_target != '' && !ctype_digit($include_target)) {
 									$include_target = empty($_SERVER['REQUEST_URI']) ? '' : $_SERVER['REQUEST_URI'];
-									if (preg_match('#/reader/api/0/stream/contents/feed/([A-Za-z0-9\'!*()%$_.~+-]+)#', $include_target, $matches)) {
+									if (preg_match('#/reader/api/0/stream/contents/feed/([A-Za-z0-9\'!*()%$_.~+-]+)#', $include_target, $matches) === false) {
```
Comment (Alkarex) [inline p/api/greader.php]: This looks wrong

File: lib/Minz/ModelPdo.php [hunk 4]
```diff
@@ -136,14 +141,13 @@ public function __construct(?string $currentUser = null, ?Minz_Pdo $currentPdo =
...
+			$db = Minz_Configuration::get('system')->db;
 
-		$db = Minz_Configuration::get('system')->db;
-
-		throw new Minz_PDOConnectionException(
+			throw new Minz_PDOConnectionException(
 				$ex->getMessage(),
 				$db['user'], Minz_Exception::ERROR
 			);
+		}
```
Comment (Alkarex) [inline lib/Minz/ModelPdo.php]: What happened here? This looks quite wrong

File: lib/Minz/Pdo.php [hunk 5]
```diff
@@ -28,7 +28,7 @@ private function autoPrefix(string $sql): string {
 	}
 
 	protected function preSql(string $statement): string {
-		if (preg_match('/^(?:UPDATE|INSERT|DELETE)/i', $statement)) {
+		if (is_int(preg_match('/^(?:UPDATE|INSERT|DELETE)/i', $statement))) {
```
Comment (Alkarex) [inline lib/Minz/Pdo.php]: That is wrong for the return case of `0`

File: lib/Minz/Request.php [hunk 6]
```diff
@@ -329,7 +331,7 @@ public static function serverIsPublic(string $address): bool {
 
 	private static function requestId(): string {
 		if (empty($_GET['rid']) || !ctype_xdigit($_GET['rid'])) {
-			$_GET['rid'] = uniqid();
+			$_GET['rid'] = uniqid('', true);
```
Comment (Alkarex) [inline lib/Minz/Request.php]: No need to waste entropy there

File: lib/lib_rss.php [hunk 7]
```diff
@@ -116,14 +116,14 @@ function checkUrl(string $url, bool $fixScheme = true) {
 	if ($url == '') {
 		return '';
 	}
-	if ($fixScheme && !preg_match('#^https?://#i', $url)) {
+	if ($fixScheme && preg_match('#^https?://#i', $url) === false) {
```
Comment (Alkarex) [inline lib/lib_rss.php]: Wrong, we mostly need to test for `0` here

File: lib/favicons.php [hunk 8]
```diff
@@ -119,6 +121,7 @@ function download_favicon(string $url, string $dest): bool {
 			}
 		}
 	}
-	return ($favicon != '' && file_put_contents($dest, $favicon)) ||
+	$filePutContent = file_put_contents($dest, $favicon);
```
Comment (Alkarex) [inline lib/favicons.php]: No, wrong order of logic: we should not write if `$favicon == ''`...

**Comment by Alkarex** [comment]
I have fixed several regressions and other smaller details

**Comment by Alkarex** [comment]
Let's not hurry making more changes of this type, especially because I believe that it leads to less readable and more error-prone code in many situations, for instance the calls to `preg_grep()`, `preg_match()`. Indeed, this original PR had several regressions, which I am guessing were at least partially due to the more complicated / less readable syntax.
I think we should relax the `booleansInConditions` rule to accept { 0, 1 } (and maybe null) in addition of { true, false }.
https://github.com/phpstan/phpstan-strict-rules/issues/221

File: lib/http-conditional.php [hunk 9]
```diff
@@ -182,7 +182,7 @@ function httpConditional(int $UnixTimeStamp, int $cacheSeconds = 0, int $cachePr
  * Reference rfc2616-sec14.html#sec14.11
  */
 function _httpConditionalCallBack(string $buffer, int $mode = 5): string {
-	if (extension_loaded('zlib') && (!ini_get('zlib.output_compression'))) {
+	if (extension_loaded('zlib') && (ini_get('zlib.output_compression') === false)) {
```
Comment (Alkarex) [inline lib/http-conditional.php]: I think this is wrong as well:
* https://php.net/function.ini-get
* https://php.net/zlib.configuration#ini.zlib.output-compression

```suggestion
	if (extension_loaded('zlib') && (ini_get('zlib.output_compression') == false)) {
```

**Comment by ColonelMoutarde** [comment]
> Let's not hurry making more changes of this type, especially because I believe that it leads to less readable and more error-prone code in many situations, for instance the calls to `preg_grep()`, `preg_match()`. Indeed, this original PR had several regressions, which I am guessing were at least partially due to the more complicated / less readable syntax. I think we should relax the `booleansInConditions` rule to accept { 0, 1 } (and maybe null) in addition of { true, false }. [phpstan/phpstan-strict-rules#221](https://github.com/phpstan/phpstan-strict-rules/issues/221)

hello, the problem is polymorphism. it must be reduced as much as possible. a function must return only one type. 

See : 
https://www.phptutorial.net/php-oop/php-polymorphism/

if a native php function is polymorphic (for historical reasons), the best thing to do is to create a method that encapsulates and returns only what is necessary.

File: lib/Minz/ModelPdo.php (hunk 4 above)
Comment (ColonelMoutarde) [reply lib/Minz/ModelPdo.php]: if I don't do this phpstan returns an error because $ex is unknown at this point

Comment (Alkarex) [reply lib/Minz/ModelPdo.php]: I have reverted without problem, as far as I can see. But the point was that the logic was broken with your changes.

**Pull request merged**